lib/transport/selenium-webdriver/appium.js

```diff
@@ -43,45 +43,41 @@ module.exports = class AppiumServer extends SeleniumServer {
   }
 
   createSessionOptions(argv) {
-    // TODO: Re-think the below temp solution
-    if (TransportFactory.usingBrowserstack(this.settings)) {
-      return super.createSessionOptions(argv);
-    }
-
-    const options = this.desiredCapabilities;
-
     // break 'appium:options' to individual configs
-    if (isObject(options['appium:options'])) {
-      const appiumOptions = options['appium:options'];
+    if (isObject(this.desiredCapabilities['appium:options'])) {
+      const appiumOptions = this.desiredCapabilities['appium:options'];
       for (let key of Object.keys(appiumOptions)) {
         const value = appiumOptions[key];
 
         if (!key.startsWith('appium:')) {
           key = `appium:${key}`;
         }
-        options[key] = value;
+        this.desiredCapabilities[key] = value;
       }
 
-      delete options['appium:options'];
+      delete this.desiredCapabilities['appium:options'];
     }
 
     // if `appium:chromedriverExecutable` is present and left blank,
     // assign the path of binary from `chromedriver` NPM package to it.
-    if (options['appium:chromedriverExecutable'] === '') {
+    if (this.desiredCapabilities['appium:chromedriverExecutable'] === '') {
       const chromedriver = this.seleniumCapabilities.getChromedriverPath();
       if (chromedriver) {
-        options['appium:chromedriverExecutable'] = chromedriver;
+        this.desiredCapabilities['appium:chromedriverExecutable'] = chromedriver;
       }
     }
 
-    return options;
+    return super.createSessionOptions(argv) || this.desiredCapabilities;
   }
 
   createDriver({options = this.desiredCapabilities} = {}) {
-    const httpClient = new http.HttpClient(this.getServerUrl());
+    // If creating a session with BrowserStack Automate, use Selenium's session builder.
+    if (TransportFactory.usingBrowserstack(this.settings) && this.productNamespace === 'automate') {
+      return super.createDriver({options});
+    }
```

Comment:
We can also add another predicate in the `if` statement to check if `options` is an instance of `Capabilities`, because we cannot use Selenium's Builder class if the browser is not supported by Selenium.